examples/rust/udp-push-pop.rs

```diff
@@ -95,173 +100,175 @@ fn freesga(libos: &mut LibOS, sga: demi_sgarray_t) {
-            Ok(qt) => qt,
-            Err(e) => {
-                // If error, close socket.
-                close(&mut libos, sockqd);
-                anyhow::bail!("pop failed: {:?}", e)
-            },
-        };
-        let sga: demi_sgarray_t = match libos.wait(qt, None) {
-            Ok(qr) if qr.qr_opcode == demi_opcode_t::DEMI_OPC_POP => unsafe { qr.qr_value.sga },
-            Ok(_) => {
-                // If error, close socket.
-                close(&mut libos, sockqd);
-                anyhow::bail!("unexpected result")
-            },
-            Err(e) => {
-                // If error, close socket.
-                close(&mut libos, sockqd);
-                anyhow::bail!("operation failed: {:?}", e)
-            },
+        match self.libos.bind(self.sockqd, local) {
+            Ok(()) => (),
+            Err(e) => anyhow::bail!("bind failed: {:?}", e),
         };
 
-        // Sanity check received data.
-        let ptr: *mut u8 = sga.sga_segs[0].sgaseg_buf as *mut u8;
-        let len: usize = sga.sga_segs[0].sgaseg_len as usize;
-        let slice: &mut [u8] = unsafe { slice::from_raw_parts_mut(ptr, len) };
-        for x in slice {
-            if *x != FILL_CHAR {
-                // If not matching, close socket and free scatter-gather array.
-                freesga(&mut libos, sga);
-                close(&mut libos, sockqd);
-                anyhow::bail!("fill check failed: expected={:?} received={:?}", FILL_CHAR, *x);
+        // Get at least nreceives.
+        for i in 0..nreceives {
+            // Receive data.
+            let qt: QToken = match self.libos.pop(self.sockqd, None) {
+                Ok(qt) => qt,
+                Err(e) => anyhow::bail!("pop failed: {:?}", e),
+            };
+            self.sga = match self.libos.wait(qt, None) {
+                Ok(qr) if qr.qr_opcode == demi_opcode_t::DEMI_OPC_POP => unsafe { Some(qr.qr_value.sga) },
+                Ok(_) => anyhow::bail!("unexpected result"),
+                Err(e) => anyhow::bail!("operation failed: {:?}", e),
+            };
+
+            // Sanity check received data.
+            let ptr: *mut u8 = self.sga.expect("should be a valid sgarray").sga_segs[0].sgaseg_buf as *mut u8;
+            let len: usize = self.sga.expect("should be a valid sgarray").sga_segs[0].sgaseg_len as usize;
+            let slice: &mut [u8] = unsafe { slice::from_raw_parts_mut(ptr, len) };
+            for x in slice {
+                demikernel::ensure_eq!(*x, fill_char);
             }
-        }
 
-        if let Err(e) = libos.sgafree(sga) {
-            // If error, close socket.
-            close(&mut libos, sockqd);
-            anyhow::bail!("failed to release scatter-gather array: {:?}", e)
+            match self.libos.sgafree(self.sga.expect("should be a valid sgarray")) {
+                Ok(_) => self.sga = None,
+                Err(e) => anyhow::bail!("failed to release scatter-gather array: {:?}", e),
```

Comment:
You may avoid these calls to `except()`, but enclosing this in a `if let Some() { .. }` statement.

Reply:
Done.